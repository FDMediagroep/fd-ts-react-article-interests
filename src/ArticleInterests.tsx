import React, { PureComponent } from "react";
import { createGlobalStyle } from "styled-components";
import TypoGraphy, { getAllTextStyles } from "@fdmg/fd-typography";
import Card, {CardTypes, CardStyle} from '@fdmg/fd-card';
import {FollowButton, FollowButtonStyle} from '@fdmg/fd-buttons';

interface Interest {
    activeButtonText?: string;
    alertSelected?: boolean;
    addInterestLink?: string;
    addAlertLink?: string;
    buttonText?: string;
    deleteInterestLink?: string;
    deleteAlertLink?: string;
    link?: string;
    selected?: boolean;
    tag: string;
    uuid: string;
}

export interface Props {
    cardStyle: CardTypes;
    className?: string;
    onDisableAlertClick: (tag: string) => void;
    onEnableAlertClick: (tag: string) => void;
    onFollowClick: (tag: string) => void;
    onUnfollowClick: (tag: string) => void;
    interests: Interest[];
    title?: string;
    titleLink?: string;
}

export default class ArticleInterests extends PureComponent<Props, any> {

    onFollowClick: React.ReactEventHandler<HTMLButtonElement> = (e) => {
        const currentTarget = e.currentTarget;
        if (currentTarget.getAttribute('data-selected') === 'true') {
            this.props.onUnfollowClick(currentTarget.getAttribute('data-tag') as string);
        } else {
            this.props.onFollowClick(currentTarget.getAttribute('data-tag') as string);
        }
    }

    onEnableAlertClick: React.ReactEventHandler<HTMLElement> = (e) => {
        const currentTarget = e.currentTarget;
        this.props.onEnableAlertClick(currentTarget.getAttribute('data-tag') as string);
    }

    onDisableAlertClick: React.ReactEventHandler<HTMLElement> = (e) => {
        const currentTarget = e.currentTarget;
        this.props.onDisableAlertClick(currentTarget.getAttribute('data-tag') as string);
    }

    render() {
        return(
            <>
                <GlobalStyle/>
                <Card cardStyle={this.props.cardStyle} className={`fd-article-interests${this.props.className ? ` ${this.props.className}` : ''}`}>
                    <TypoGraphy className="fd-article-interests-h" textStyle="card-h">
                        <h3><a href={`${this.props.titleLink ? this.props.titleLink : '/mijn-nieuws'}`}>{this.props.title ? this.props.title : 'Volgen via mijn nieuws'}</a></h3>
                    </TypoGraphy>
                    <ul>
                        {
                            this.props.interests.map((interest: Interest) => (
                                <li key={interest.uuid}>
                                    <div className={`interest-container${interest.selected ? ' selected' : ''}`}>
                                        <a href={`${interest.link ? interest.link : `/tag/${interest.tag}`}`} title={interest.tag}>{interest.tag}</a>
                                        <span className="interest-controls">
                                            <span className={`alert${interest.alertSelected ? ' selected' : ''}`} data-addurl={interest.addAlertLink ? interest.addAlertLink : '/add-alert'} data-deleteurl={interest.addAlertLink ? interest.addAlertLink : '/add-alert'}>
                                                <i className="default-icon icon-bell" data-tag={interest.tag} onClick={this.onEnableAlertClick}/>
                                                <i className="active-icon icon-bell1" data-tag={interest.tag} onClick={this.onDisableAlertClick}/>
                                            </span>
                                            <FollowButton
                                                buttonStyle={this.props.cardStyle === 'persoonlijk' ? this.props.cardStyle : 'default'}
                                                onClick={this.onFollowClick}
                                                tag={interest.tag}
                                                selected={interest.selected}
                                                followLink={interest.addInterestLink}
                                                unFollowLink={interest.deleteInterestLink}
                                                followButtonText={interest.buttonText}
                                                unFollowButtonText={interest.activeButtonText}
                                            />
                                        </span>
                                    </div>
                                </li>
                                )
                            )
                        }
                    </ul>
                </Card>
            </>
        );
    }
}

const GlobalStyle = createGlobalStyle`
.fd-article-interests {
    padding: 15px 15px 0 15px;
    h3.fd-article-interests-h {
        min-height: 20px;
        margin-bottom: 15px;
    }
    h3.fd-article-interests-h a {
        color: inherit;
        text-decoration: inherit;
    }

    ul {
        padding: 0;
        margin: 0 -15px;
        list-style: none;
    }

    li {
        display: inline-block;
        width: 100%;
        border-top: 1px solid rgba(0, 0, 0, 0.1);
        > div {
            margin: 10px;
            display: flex;
            position: relative;
            height: 30px;
        }
    }

    .interest-container a {
        flex-grow: 1;
        position: relative;
        top: 5px;
        color: #191919;
        &:hover {
            color: #49a4a2;
        }
        display: block;
        white-space: nowrap;
        text-overflow: ellipsis;
        overflow: hidden;
        text-decoration: none;
        font-family: 'ProximaNovaRegular', Helvetica, sans-serif;
    }

    .interest-controls {
        display: flex;
        justify-content: flex-end;
        flex: 0 0 140px;

        .alert {
            margin-right: .5rem;
            color: #677381;
            display: flex;
            flex-direction: column;
            align-items: center;
            justify-content: center;
            i {
                display: none;
                cursor: pointer;
            }
        }
    }

    .interest-container.selected {
        .default-icon {
            display: inline;
        }
    }

    .alert.selected {
        .default-icon {
            display: none;
        }
        .active-icon {
            display: inline;
        }
    }

    &.persoonlijk {
        h3 {
            color: #e57e30;
        }

        .interest-container a:hover {
            color: #e57e30;
        }
    }
}
`;

export const ArticleInterestsStyle = createGlobalStyle`
${(CardStyle as any).globalStyle.rules}
${getAllTextStyles(['card-h']).globalStyle.rules}
${(FollowButtonStyle as any).globalStyle.rules}
${(GlobalStyle as any).globalStyle.rules}
`;
