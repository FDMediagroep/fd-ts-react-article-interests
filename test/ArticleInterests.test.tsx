import TestRenderer from 'react-test-renderer';
import ArticleInterests from '../src/ArticleInterests';
import React from 'react';
import uniqid from 'uniqid';

const onDisableAlertClick = () => {};
const onEnableAlertClick = () => {};
const onFollowClick = () => {};
const onUnfollowClick = () => {};
const interests = [{
    tag: 'Detailhandel',
    uuid: uniqid()
}, {
    selected: true,
    tag: 'Eten & Drinken',
    uuid: uniqid()
}, {
    alertSelected: true,
    selected: true,
    tag: 'Supermarkt',
    uuid: uniqid()
}];

describe('ArticleInterests', () => {
    test('renders correctly', () => {
        let articleInterests = TestRenderer.create(<ArticleInterests
            cardStyle="article"
            onDisableAlertClick={onDisableAlertClick}
            onEnableAlertClick={onEnableAlertClick}
            onFollowClick={onFollowClick}
            onUnfollowClick={onUnfollowClick}
            title="Volgen via mijn nieuws"
            titleLink="https://fd.nl/mijn-nieuws"
            interests={interests}
        />);
        expect(articleInterests.toJSON()).toMatchSnapshot();

        articleInterests = TestRenderer.create(<ArticleInterests
            cardStyle="persoonlijk"
            onDisableAlertClick={onDisableAlertClick}
            onEnableAlertClick={onEnableAlertClick}
            onFollowClick={onFollowClick}
            onUnfollowClick={onUnfollowClick}
            title="Volgen via mijn nieuws"
            titleLink="https://fd.nl/mijn-nieuws"
            interests={interests}
        />);
        expect(articleInterests.toJSON()).toMatchSnapshot();
    });

});
