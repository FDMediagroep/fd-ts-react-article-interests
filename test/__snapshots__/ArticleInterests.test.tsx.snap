// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleInterests renders correctly 1`] = `
<section
  className="fd-article-interests"
>
  <h3
    className="fd-typography card-h fd-article-interests-h"
    textStyle="card-h"
  >
    <a
      href="https://fd.nl/mijn-nieuws"
    >
      Volgen via mijn nieuws
    </a>
  </h3>
  <ul>
    <li>
      <div
        className="interest-container"
      >
        <a
          href="/tag/Detailhandel"
          title="Detailhandel"
        >
          Detailhandel
        </a>
        <span
          className="interest-controls"
        >
          <span
            className="alert"
            data-addurl="/add-alert"
            data-deleteurl="/add-alert"
          >
            <i
              className="default-icon icon-bell"
              data-tag="Detailhandel"
              onClick={[Function]}
            />
            <i
              className="active-icon icon-bell1"
              data-tag="Detailhandel"
              onClick={[Function]}
            />
          </span>
          <button
            data-addurl="/add-interest"
            data-deleteurl="/delete-interest"
            data-tag="Detailhandel"
            onClick={[Function]}
          >
            <span
              className="cross"
            />
             
            <span
              className="button-text default-text"
            >
              Volg
            </span>
            <span
              className="button-text active-text"
            >
              Ontvolg
            </span>
          </button>
        </span>
      </div>
    </li>
    <li>
      <div
        className="interest-container selected"
      >
        <a
          href="/tag/Eten & Drinken"
          title="Eten & Drinken"
        >
          Eten & Drinken
        </a>
        <span
          className="interest-controls"
        >
          <span
            className="alert"
            data-addurl="/add-alert"
            data-deleteurl="/add-alert"
          >
            <i
              className="default-icon icon-bell"
              data-tag="Eten & Drinken"
              onClick={[Function]}
            />
            <i
              className="active-icon icon-bell1"
              data-tag="Eten & Drinken"
              onClick={[Function]}
            />
          </span>
          <button
            data-addurl="/add-interest"
            data-deleteurl="/delete-interest"
            data-selected={true}
            data-tag="Eten & Drinken"
            onClick={[Function]}
          >
            <span
              className="cross"
            />
             
            <span
              className="button-text default-text"
            >
              Volg
            </span>
            <span
              className="button-text active-text"
            >
              Ontvolg
            </span>
          </button>
        </span>
      </div>
    </li>
    <li>
      <div
        className="interest-container selected"
      >
        <a
          href="/tag/Supermarkt"
          title="Supermarkt"
        >
          Supermarkt
        </a>
        <span
          className="interest-controls"
        >
          <span
            className="alert selected"
            data-addurl="/add-alert"
            data-deleteurl="/add-alert"
          >
            <i
              className="default-icon icon-bell"
              data-tag="Supermarkt"
              onClick={[Function]}
            />
            <i
              className="active-icon icon-bell1"
              data-tag="Supermarkt"
              onClick={[Function]}
            />
          </span>
          <button
            data-addurl="/add-interest"
            data-deleteurl="/delete-interest"
            data-selected={true}
            data-tag="Supermarkt"
            onClick={[Function]}
          >
            <span
              className="cross"
            />
             
            <span
              className="button-text default-text"
            >
              Volg
            </span>
            <span
              className="button-text active-text"
            >
              Ontvolg
            </span>
          </button>
        </span>
      </div>
    </li>
  </ul>
</section>
`;
